File: python/pager.py
import os
import math
import constants

class Pager:

  def __init__(self, filename):
    mode = 'r+b' if os.path.exists(filename) else 'w+b'
    self.file = open(filename, mode)
    self.file_length = os.stat(filename).st_size
    self.num_pages = math.ceil(self.file_length/constants.PAGE_SIZE)
    self.pages = [None] * constants.TABLE_MAX_PAGES

  def close(self):
    self.file.close()

  def load_or_create_page(self, page_num):
    # get number of full and incomplete pages in the file

    # if there is incomplete page data in the file, read it into page
    if (page_num <= self.num_pages):
      self.file.seek(page_num * constants.PAGE_SIZE)
      return bytearray(self.file.read(constants.PAGE_SIZE))

    return bytearray()

  def get_page(self, page_num):
    # make sure that the page isn't out of bounds
    if (page_num > constants.TABLE_MAX_PAGES):
      exit('Tried to fetch page number out of bounds. ' + page_num + ' > ' + constants.TABLE_MAX_PAGES)

    # if the page isn't already loaded up, get it from disk (or make a blank one)
    if (self.pages[page_num] == None):
      self.pages[page_num] = self.load_or_create_page(page_num)

    return self.pages[page_num]

  def overwrite_page(self, node, page_num):
    self.pages[page_num] = node.serialize()


  def insert_new_page(self, node):
    node.serialize()
    page_num = self.get_unused_page_num()
    self.pages[page_num] = node.page
    self.num_pages += 1

    return page_num


  def flush_page(self, page_num):
    page = self.get_page(page_num)
    if (page == None):
      exit('Tried to flush null page')

    self.file.seek(page_num * constants.PAGE_SIZE)
    self.file.write(page)

  def get_unused_page_num(self):
    return self.num_pages

File: c/pager.c
#include <fcntl.h>
#include <errno.h>
#include <unistd.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "pager.h"

void* create_page(Pager* pager, uint32_t page_num) {
  void* page = malloc(PAGE_SIZE);
  uint32_t num_pages = pager->file_length / PAGE_SIZE;

  // We might save a partial page at the end of the file
  if (pager->file_length % PAGE_SIZE) {
    num_pages += 1;
  }

  // if we're missing some page information, read data into the page from the file
  if (page_num <= num_pages) {
    lseek(pager->file_descriptor, page_num * PAGE_SIZE, SEEK_SET);
    ssize_t bytes_read = read(pager->file_descriptor, page, PAGE_SIZE);
    if (bytes_read == -1) {
      printf("Error reading file: %d\n", errno);
      exit(EXIT_FAILURE);
    }
  }

  return page;    
}

void* get_page(Pager* pager, uint32_t page_num) {
  // if the page number is out of bounds, error & exit
  if (page_num > TABLE_MAX_PAGES) {
    printf("Tried to fetch page number out of bounds. %d > %d\n", page_num, TABLE_MAX_PAGES);
    exit(EXIT_FAILURE);
  }

  // if the requested page doesn't exist already, make one
  if (pager->pages[page_num] == NULL) {
    pager->pages[page_num] = create_page(pager, page_num);
  }

  return pager->pages[page_num];
}

Pager* pager_open(const char* filename) {
  int fd = open(
    filename,
    O_RDWR  | O_CREAT,    // Read/Write mode, create file if it does not exist
    S_IWUSR | S_IRUSR     // User write permission, or user read permission
  );

  if (fd == -1) {
    printf("Unable to open file\n");
    exit(EXIT_FAILURE);
  }

  off_t file_length = lseek(fd, 0, SEEK_END);

  Pager* pager = malloc(sizeof(Pager));
  pager->file_descriptor = fd;
  pager->file_length = file_length;

  for (uint32_t i = 0; i < TABLE_MAX_PAGES; i++) {
    pager->pages[i] = NULL;
  }

  return pager;
}

void pager_flush(Pager* pager, uint32_t page_num, uint32_t size) {
  if (pager->pages[page_num] == NULL) {
    printf("Tried to flush null page\n");
    exit(EXIT_FAILURE);
  }

  off_t offset = lseek(pager->file_descriptor, page_num * PAGE_SIZE, SEEK_SET);

  if (offset == -1) {
    printf("Error seeking: %d\n", errno);
    exit(EXIT_FAILURE);
  }

  ssize_t bytes_written =
      write(pager->file_descriptor, pager->pages[page_num], size);

  if (bytes_written == -1) {
    printf("Error writing: %d\n", errno);
    exit(EXIT_FAILURE);
  }
}
 